PaintTest/main.c

```diff
@@ -0,0 +1,691 @@
+	if (expected.paint.bound_box_length.x != actual.paint.bound_box_length.x) return false;
+	if (expected.paint.bound_box_length.y != actual.paint.bound_box_length.y) return false;
+	if (expected.paint.bound_box_length.z != actual.paint.bound_box_length.z) return false;
+	if (function != PAINT_98196C) {
+		if (expected.paint.bound_box_offset.x != actual.paint.bound_box_offset.x) return false;
+		if (expected.paint.bound_box_offset.y != actual.paint.bound_box_offset.y) return false;
+		if (expected.paint.bound_box_offset.z != actual.paint.bound_box_offset.z) return false;
+	}
+	if (expected.paint.z_offset != actual.paint.z_offset) return false;
+	if (expected.paint.rotation != actual.paint.rotation) return false;
+
+	return true;
+}
+
+static bool assertFunctionCallArrayEquals(function_call expected[], uint8 expectedCount, function_call actual[], uint8 actualCount) {
+	if (expectedCount != actualCount) {
+		return false;
+	}
+
+	for (int i = 0; i < expectedCount; i++) {
+		function_call expectedCall = expected[i];
+		function_call actualCall = actual[i];
+
+		if (!assertFunctionCallEquals(expectedCall, actualCall)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void printImageId(uint32 input, utf8string *out) {
+	uint32 image = input & 0x7FFFF;
+	uint32 palette = input & ~0x7FFFF;
+
+	bool allocated = false;
+	utf8string paletteName;
+	if (palette == PALETTE_98)paletteName = "PALETTE_98";
+	else if (palette == PALETTE_9C)paletteName = "PALETTE_9C";
+	else if (palette == PALETTE_A0)paletteName = "PALETTE_A0";
+	else if (palette == PALETTE_A4)paletteName = "PALETTE_A4";
+	else {
+		paletteName = malloc(16);
+		sprintf(paletteName, "0x%08X", palette);
+	}
+
+	if (image == 0) {
+		sprintf(*out, "%s", paletteName);
+	} else if (image & 0x70000) {
+		sprintf(*out, "%s | vehicle.base_image_id + %d", paletteName, image & ~0x70000);
+	} else {
+		sprintf(*out, "%s | %d", paletteName, image);
+	}
+}
+
+static void printFunctionCall(utf8string *out, function_call call) {
+	utf8string imageId = malloc(64);
+	printImageId(call.supports.colour_flags, &imageId);
+	switch (call.function) {
+	case SUPPORTS_WOOD_A:
+		sprintf(*out, "wooden_a_supports_paint_setup(%d, %d, %d, %s)", call.supports.type, call.supports.special, call.supports.height, imageId);
+		return;
+	case SUPPORTS_WOOD_B:
+		sprintf(*out, "wooden_a_supports_paint_setup(%d, %d, %d, %s)", call.supports.type, call.supports.special, call.supports.height, imageId);
+		return;
+
+	case SUPPORTS_METAL_A:
+		sprintf(*out, "metal_a_supports_paint_setup(%d, %d, %d, %d, %s)", call.supports.type, call.supports.segment, call.supports.special, call.supports.height, imageId);
+		return;
+	case SUPPORTS_METAL_B:
+		sprintf(*out, "metal_b_supports_paint_setup(%d, %d, %d, %d, %s)", call.supports.type, call.supports.segment, call.supports.special, call.supports.height, imageId);
+		return;
+	}
+
+	utf8string name;
+	switch (call.function) {
```

Comment:
Perhaps provide a default option to squash the overzealous warning about potentially uninitialised pointer.